### PR TITLE
Tooltip refactor

### PR DESCRIPTION
**Pull request by Zarel**
(I just wanted to try out draft pull requests.)

This is what I've been working on for the client refactor for the past entire month.

I honestly got _really_ burnt out by this.

- [x] Tooltips work in replays now
- [x] Calculation is better (correctly handles more corner cases)
- [x] Explanations are better (better messages for Magic Room etc)
- [x] Tooltips for sidebar pokemon
- [x] Tooltips can be locked
- [x] You can copy/paste from locked tooltips

### CODE REVIEW

**Comment by Zarel** [comment]
I think this is basically at parity now. It was a horrible slog but it's done now. The last part was surprisingly fast for how slow the rest was.

**Comment by Zarel** [comment]
The Travis error is from `@types/node`, but this commit _removes_ the Node dependency, so idk what it's complaining about. :/

**Comment by Zarel** [comment]
This pull request **introduces 5 alerts** when merging 3dd1afd9c037caa12ad7deef30913bc1842611e8 into 1e39effb5998cbb0fbff1a0dd91a8129c98edd01 - [view on LGTM.com](https://lgtm.com/projects/g/Zarel/Pokemon-Showdown-Client/rev/pr-1e573c074aa9209351e847c287dfd1223967ab68)

**new alerts:**

* 2 for Unused variable, import, function or class
* 2 for Useless assignment to property
* 1 for Useless assignment to local variable

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by Zarel** [comment]
This pull request **introduces 2 alerts** when merging 46e0debe4b226e74ab0b1025cb2ab768a85679a8 into 1e39effb5998cbb0fbff1a0dd91a8129c98edd01 - [view on LGTM.com](https://lgtm.com/projects/g/Zarel/Pokemon-Showdown-Client/rev/pr-bdbdd41936b99c362caa72a1665cd3df8f9d0227)

**new alerts:**

* 2 for Useless assignment to property

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by Zarel** [comment]
This pull request **introduces 2 alerts** when merging 691b286919346147a8685ed7821d168d6733270e into b0b014fdda0defeefdbfa94704e3c7bf5b3758e6 - [view on LGTM.com](https://lgtm.com/projects/g/Zarel/Pokemon-Showdown-Client/rev/pr-099871a3c9be62f2391b648697eb8782cbfd4226)

**new alerts:**

* 2 for Useless assignment to property

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by Zarel** [comment]
This pull request **introduces 2 alerts** when merging fd39a8d6651741184059e41c6549069f853bd7e0 into 2b1400631d9126da40f24ce0769eee2344ab44d7 - [view on LGTM.com](https://lgtm.com/projects/g/Zarel/Pokemon-Showdown-Client/rev/pr-0a0bfbd7d798ffc01d3a0a5513208fa7c3aebebd)

**new alerts:**

* 2 for Useless assignment to property

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: src/battle-tooltips.ts [hunk 1]
```diff
@@ -0,0 +1,1514 @@
...
+			value.abilityModify(this.battle.gen > 6 ? 1.2 : 1.3, "Pixilate");
+			value.abilityModify(this.battle.gen > 6 ? 1.2 : 1.3, "Refrigerate");
+		}
+		if (move.flags['punch']) {
+			value.abilityModify(1.2, 'Iron Fist');
+		}
+		if (move.recoil || move.hasCustomRecoil) {
+			value.abilityModify(1.2, 'Reckless');
+		}
+		if (this.battle.gen > 6) {
```
Comment (MacChaeger) [inline src/battle-tooltips.ts]: I believe it was established that Normalize doesn’t boost moves that change type on their own, so this should be grouped with the -ates.

Comment (MacChaeger) [reply src/battle-tooltips.ts]: Er, it should at least check `noTypeOverride`

File: src/battle-tooltips.ts [hunk 2]
```diff
@@ -0,0 +1,1514 @@
...
+		if (this.tryAbility('Klutz') && !ignoreKlutz.includes(itemName)) {
+			this.comment.push(` (${itemName} suppressed by Klutz)`);
+			return false;
+		}
+		return true;
+	}
+	tryAbility(abilityName: string) {
+		if (abilityName !== this.abilityName) return false;
+		if (this.pokemon && this.pokemon.volatiles['gastroacid']) {
+			this.comment.push(` (${abilityName} suppressed by Gastro Acid)`);
```
[the remark this answers is not in the document]
Comment (Zarel) [reply src/battle-tooltips.ts]: Core Enforcer sends the same volatile as Gastro Acid.

File: src/battle-tooltips.ts (hunk 1 above)
Comment (Zarel) [reply src/battle-tooltips.ts]: For the record, this is totally not in the part of the code that I changed...

**Comment by Zarel** [comment]
This pull request **introduces 2 alerts** when merging 8b8de1f8f10d36d509f2a0ce34dc362a0cafe420 into 2b1400631d9126da40f24ce0769eee2344ab44d7 - [view on LGTM.com](https://lgtm.com/projects/g/Zarel/Pokemon-Showdown-Client/rev/pr-ca38bba30a56e8f46d8c7325f9161f505473142f)

**new alerts:**

* 2 for Useless assignment to property

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: src/battle.ts [hunk 3]
```diff
@@ -452,29 +452,58 @@ class Pokemon {
...
+		if (battle.hasPseudoWeather('Gravity')) {
+			return true;
+		} else if (this.volatiles['ingrain'] && battle.gen >= 4) {
+			return true;
+		} else if (this.volatiles['smackdown']) {
+			return true;
+		}
+
+		let item = toId(serverPokemon ? serverPokemon.item : this.item);
+		if (battle.hasPseudoWeather('Magic Room') || this.volatiles['embargo']) {
```
Comment (Marty-D) [inline src/battle.ts]: Klutz?

Comment (Zarel) [reply src/battle.ts]: Fair

**Comment by Zarel** [comment]
This pull request **introduces 2 alerts** when merging 434afefeaa6af9c00a8cef02c38ee8015fa51dc7 into 2b1400631d9126da40f24ce0769eee2344ab44d7 - [view on LGTM.com](https://lgtm.com/projects/g/Zarel/Pokemon-Showdown-Client/rev/pr-9566b6bebb86c1015701d676fb432489f1b18d5b)

**new alerts:**

* 2 for Useless assignment to property

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Pull request merged**